File: Python/2096_내려가기.py
import sys

if __name__ == '__main__':
	N = int(sys.stdin.readline())
	
	pre_score = [[0]*2 for _ in range(3)] # 이전까지의 최대 최소 score 담겨 있음
	score = [[0]*2 for _ in range(3)] # 현재까지의 최대 최소 score 담겨 있음
	
	arr = list(map(int,sys.stdin.readline().split()))
	for i in range(3):
		score[i] = [arr[i], arr[i]]
	pre_score = [[x1, x2] for x1, x2 in score] # 초기화
	
	for _ in range(1,N):
		arr = list(map(int,sys.stdin.readline().split()))
		for i in range(3):
			if i-1 >=0:
				if i+1 < 3: # 2열: 이전 scroe 1,2,3 열 중 최대, 최솟값 이용하여 계산
					score[i][0] = min(pre_score[i][0], pre_score[i+1][0], pre_score[i-1][0]) + arr[i]
					score[i][1] = max(pre_score[i][1], pre_score[i+1][1], pre_score[i-1][1]) + arr[i]
				else: # 3열: 이전 scroe 2,3 열 중 최대, 최솟값 이용하여 계산
					score[i][0] = min(pre_score[i][0], pre_score[i-1][0]) + arr[i]
					score[i][1] = max(pre_score[i][1], pre_score[i-1][1]) + arr[i]
					
			else:
				if i+1 < 3: # 1열: 이전 scroe 1,2 열 중 최대, 최솟값 이용하여 계산
					score[i][0] = min(pre_score[i][0], pre_score[i+1][0]) + arr[i]
					score[i][1] = max(pre_score[i][1], pre_score[i+1][1]) + arr[i]
		pre_score = [[x1, x2] for x1, x2 in score] # 다음 계산을 위해 pre_score 갱신
	
	s_min = float('inf')
	s_max = 0
	for i in range(3): # 구한 최대 최솟값들 중 최대 최소 구하기
		s_min = min(s_min,score[i][0])
		s_max = max(s_max,score[i][1])
		
	print("%d %d" %(s_max, s_min))

File: C++/2096_내려가기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Score{
	int s_min, s_max;
};

int main() {
	int N;
	cin >> N;
	
	vector<Score> score(3);
	vector<Score> pre_score(3);
	for(int i = 0; i < 3; i++){
		int tmp;
		cin >> tmp;
		score[i] = {tmp,tmp};
		pre_score[i] = {tmp,tmp};
	} // 점수의 {최소, 최대}, pre_score는 이전 행까지의 {최소, 최대}를 의미 score는 현재 행까지의{최소, 최대}를 의미 / 초기화
	
	
	for(int i = 1; i < N; i++){
		for(int j = 0; j < 3; j++){
			int tmp;
			cin >> tmp;
			if(j-1 >= 0){
				if(j+1 < 3){ // 2번째 열의 경우: pre_score 1,2,3열 중 가장 큰 값, 작은 값 찾아서 최대, 최소에 넣어주되 현재의 값을 더해서 넣어야 함
					int s_tmp = min(pre_score[j-1].s_min, pre_score[j].s_min);
					s_tmp = min(s_tmp,pre_score[j+1].s_min);
					score[j].s_min = s_tmp + tmp;
					
					s_tmp = max(pre_score[j-1].s_max, pre_score[j].s_max);
					s_tmp = max(s_tmp, pre_score[j+1].s_max);
					score[j].s_max = s_tmp + tmp;
				}
				else{ // 3번째 열의 경우: pre_score의 2,3열 중 가장 큰 값, 작은 값 찾아서 최대, 최소에 넣어주되 현재의 값을 더해서 넣어야 함
					score[j].s_min = min(pre_score[j-1].s_min, pre_score[j].s_min) + tmp;
					score[j].s_max = max(pre_score[j-1].s_max, pre_score[j].s_max) + tmp;
				}
			}
			else{
				if(j+1 < 3){ // 1번째 열의 경우: pre_score의 1,2열 중 가장 큰 값, 작은 값 찾아서 최대, 최소에 넣어주되 현재의 값을 더해서 넣어야 함
					score[j].s_min = min(pre_score[j].s_min, pre_score[j+1].s_min) + tmp;
					score[j].s_max = max(pre_score[j].s_max, pre_score[j+1].s_max) + tmp;
				}
			}
		}
		for(int i = 0; i < 3; i++){
			pre_score[i] = score[i]; // pre_score 값 갱신(다음 계산을 위해)
		}
	}
	
	int s_min = 1000000, s_max = 0;
	for(int i = 0; i < 3; i++){ // 진짜 최대 최소 구하기 위해서는 계산 값의 최소 중 가장 최소, 최대 중 가장 최대인 값을 구해야 함
		if(s_min > score[i].s_min){
			s_min = score[i].s_min;
		}
		if(s_max < score[i].s_max){
			s_max = score[i].s_max;
		}
	}
	
	cout << s_max << " " << s_min;
	
	return 0;
}
